MAINT Codespell configuration

DimitriPapadopoulos opened the pull request:
#### What does this implement/fix? Explain your changes.

Add [codespell](https://github.com/codespell-project/codespell) to CI and fix existing typos.

#### Any other comments?

Not sure where to put file `codespell_ignore_words.txt `. Can you help?

Finally, should this really go into the Changelog?

DimitriPapadopoulos committed:
Configure codespell optimally

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

DimitriPapadopoulos commented:
I believe the linter issues predate my own changes (probably an effect of dos2unix).

glemaitre commented:
We used `black` only on the code source and ignore the examples and tutorial up to now. In addition, we run `flake8` only on the diff of each PR. If the spelling was inside of the tutorial, and this tutorial is not PEP8 compliant, then it would be logical that an error is raised. Looking at a couple of the failures, I could spot such errors. It might be worth making these files PEP8 compliant just for this PR (only the one modified).

Regarding the usage of `codespell` in the project, I think that it could be good. I personally use something similar locally in my vs code and it could be nice to have fewer typos in the project.

any thoughts @adrinjalali @thomasjpfan @NicolasHug?

DimitriPapadopoulos commented:
Ah right, all linter errors occur in `doc/tutorial/machine_learning_map/parse_path.py` as far as I can see. This one of the files that went through `dos2unix`. Should I run flake8 on this file and fix it?

NicolasHug commented:
I'm generally happy with fixing and preventing typos, my only concern is the potential risk of false positives. Did you observe a lot of these in your local setup @glemaitre ?

As a technical detail on this PR, I feel like we should fix the CRLF issues in a separate PR. 6K+ lines of change is scary for us reviewers :p !

DimitriPapadopoulos commented:
I'll move the `dos2unix` commit to a separate PR.

DimitriPapadopoulos commented:
False positives I have seen myself:
* a few variables names
* a couple _codespell_ bugs (`complies => compiles` or `theses => thesis`)
* names of contributors or names from the bibliography

glemaitre commented:
> I'm generally happy with fixing and preventing typos, my only concern is the potential risk of false positives. Did you observe a lot of these in your local setup @glemaitre ?

Locally I don't recall any false positives (or so few of them that I don't notice them) but I don't know if this [extension](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) uses the same engine.

thomasjpfan commented on .github/workflows/codespell.yml:
```diff
@@ -0,0 +1,20 @@
+      - uses: actions/checkout@v2
+      - uses: codespell-project/actions-codespell@master
+        with:
+          skip: js,v*.rst,_stop_words.py
+          ignore_words_file: .github/codespell_ignore_words.txt
```
Can this be placed in a RC file so it is easier to run locally? (https://github.com/codespell-project/codespell#using-a-config-file)

DimitriPapadopoulos replied:
Good idea, in Python projects I can put these in `setup.cfg`.

thomasjpfan reviewed (commented):
I am happy with spell checking in the CI.

To complete the PR, I think we need to add something in the contributor guide to describe how to run this locally.

DimitriPapadopoulos commented on .github/codespell_ignore_words.txt:
```diff
@@ -0,0 +1,38 @@
+aline
+amoungst
+ba
```
`Aline` is a person's name from bibliography.
`amoungst` is a valid variant of `amongst` from `sklearn/feature_extraction/_stop_words.py`
`ba` is a variable name

thomasjpfan replied:
> amoungst is a valid variant of amongst from sklearn/feature_extraction/_stop_words.py

Is this safe to remove from the ignore list now that we ignore `_stop_words.py`?

Some of these variable names are not the best. As a followup, we can start looking at the ignore list and try to remove some and create better variable names for them.

DimitriPapadopoulos replied:
Yes! I will remove it from the list.

DimitriPapadopoulos replied:
Not certain `ba` is an actual variable name. It might be a PDF command used when generating docs, or something similar you cannot change. I'd have to run codespell without ignoring `ba` to tell you. Most variable names are actually (very) well chosen.

DimitriPapadopoulos replied:
I have removed `amoungst`. Note that the `complies`false positive might be the result of an actual codespell bug or new feature (codespell-project/codespell#2062).

DimitriPapadopoulos replied:
`ba` is indeed a variable name is some cases, but it is also part of an URL!
`http://ru.wikipedia.org/wiki/%D0%92%D0%B8%D0%`**`BA`**`%D0%B8%D0%BF%D0%B5%D0%B4%D0%B8%D1%8F`

While codespell does have an `--uri-ignore-words-list` option, it has not been propagated to the GitHub action. I have opened another ticket for that (codespell-project/actions-codespell#36).

ogrisel commented on sklearn/datasets/tests/test_openml.py:
```diff
@@ -745,12 +745,12 @@ def test_fetch_openml_emotions_pandas(monkeypatch):
         "quiet.still",
         "sad.lonely",
-        "angry.aggresive",
+        "angry.aggressive",
```
This is a problem: those typos are part of the original dataset hosted at openml.org. We should not fix those and we should find a way to silence false positives for this specific file without adding typos in the global `codespell_ignore_words.txt` file. Would it be possible to tell codespell to ignore some words in a specific file?

DimitriPapadopoulos replied:
Reverted to the openml.org typos, `aggresive` and `suprised` added to the list of false positives.

ogrisel reviewed (commented):
On top of https://github.com/scikit-learn/scikit-learn/pull/21051/files#r709263298 there is also the following:

ogrisel commented:
If the false positive management and maintenance overhead to have to manage an ignore list is judged too much trouble to be worth it, I would be in favor of opening a dedicated ,one-shot PR that just fixes the currently identified typos, without the extra CI harness. Merging that one should not be controversial ;)

glemaitre commented:
+1 with Olivier proposal.

On Thu, 16 Sept 2021 at 15:29, Olivier Grisel ***@***.***>
wrote:

> If the false positive management and maintenance overhead to have to
> manage an ignore list is judged too much trouble to be worth it, I would be
> in favor of opening a dedicated ,one-shot PR that just fixes the currently
> identified typos, without the extra CI harness. Merging that one should not
> be controversial ;)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/21051#issuecomment-920903690>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABY32P6Y7TFJSD2AZ5NQMADUCHWLPANCNFSM5ECEGMFA>
> .
>


-- 
Guillaume Lemaitre
Scikit-learn @ Inria Foundation
https://glemaitre.github.io/

NicolasHug commented:
Maybe we can still keep the ignore_word list etc and make this a step in the release process?

DimitriPapadopoulos commented:
I believe the management of false positives in **`.github/codespell_ignore_words.txt`** is a small overhead. If you have a look at the list, it's pretty short. It won't happen that often.

Fixing the actual typos might be more annoying, it **will** happen more often - compare the number of words of `codespell_ignore_words.txt` with the number of typos fixed in this PR. Will all developers accept to have to fix their typos before the patch can be merged?

An alternative would be to show the results of codespell but not block the merge. It would be up to reviewers to let typos through or decide they're false positives, and perhaps update the list of false positives in `codespell_ignore_words.txt` later on, in batches. I don't know how feasible this with the current GitHub action.

DimitriPapadopoulos commented:
How to make that a step in the release process? Are the steps just documented or actually programmed?

In any case, I can split this in two PRs:
* this PR to just fix current typos
* another PR to integrate codespell into CI or the release process

DimitriPapadopoulos commented:
I have moved the actual typo fixes to #21069. It would be great if someone could merge that PR soon.

Then I can rebase this PR and use it to continue discussing the machinery to prevent future typos in this PR, either as part of CI, or as a step in the release process.

adrinjalali commented:
I like the other PR fixing the typos, but I'm also wary of this being a part of the CI, especially with the way we need to globally ignore things. I'd be in favor of having this done periodically, at every release or once a year or so, rather than being a part of CI.

DimitriPapadopoulos commented:
I'll remove the GitHub action `.github/workflows/codespell.yml`.

Is it OK to keep the codespell parameters in `setup.cfg`?
```
[codespell]
skip = js,v*.rst,_stop_words.py
ignore-words = .github/codespell_ignore_words.txt
```

Since this won't be a GitHub action, I should probably move `.github/codespell_ignore_words.txt` elsewhere. Where? Perhaps a hidden file in the root directory such as `.codespell_ignore_words.txt`?

ogrisel commented:
> Since this won't be a GitHub action, I should probably move .github/codespell_ignore_words.txt elsewhere. Where? Perhaps a hidden file in the root directory such as .codespell_ignore_words.txt?

I would put that under `build_tools`.

thomasjpfan commented on setup.cfg:
```diff
@@ -71,3 +71,7 @@ ignore =
     sklearn/utils/_weight_vector.pxd
+
+[codespell]
+skip = js,v*.rst,_stop_words.py
```
There are some auto generated files that would good to skip:

```suggestion
skip = *.js,./sklearn/feature_extraction/_stop_words.py,./doc/_build,./doc/auto_examples,./doc/modules/generated
```

I think expanding `_stop_words.py` to the full path makes it more explicit. I see that `v*.rst` is skipped because of names, but I think having spell check enabled for the changelog is a net win.

DimitriPapadopoulos replied:
I agree about the full paths, but I'd rather wait for https://github.com/codespell-project/codespell/pull/2058 to be taken into account.

About `v*.rst`, I've left these out because many projects don't want to modify the changelog, they want typos in there to remain as they are forever. I believe this is inspired by the GNU documentation on [Change Logs](https://www.gnu.org/prep/standards/html_node/Change-Logs.html). On the other hand, not everyone agrees with the GNU documentation, after all the changelog is viewed as a detailed dump of the VCS, which may appear a bit too detailed and obsolete these days. If many of you want me to fix the typos in the changelog, I'm happy to do that. However, I suggest this happens in a distinct PR, like #21069 for the rest of the typos.

DimitriPapadopoulos replied:
Also, it's `js` and not `*.js`, to exclude vendored JS scripts under `doc/themes/scikit-learn-modern/static/js`, but not other JS scripts (if any in the future).

thomasjpfan reviewed (commented):
Running `codespell` locally is great. Thank you for working on these spelling errors @DimitriPapadopoulos !

thomasjpfan commented:
When I run `codespell` locally, I see that `codespell` is looking at my `.git/logs` folder.  @DimitriPapadopoulos Are you seeing the same behavior?

DimitriPapadopoulos commented:
Indeed, that's because I had set it up for automated scans running on a git archive extraction, so without a **.git** directory. Fixed now by adding **`./.git`** to the list of files/directories to skip.

I have also added a new **`build_tools/codespell_exclude.txt`** file to exclude whole lines instead of words, specifically for names in **`v*.rst`**.

thomasjpfan commented on setup.cfg:
```diff
@@ -71,3 +71,8 @@ ignore =
     sklearn/utils/_weight_vector.pxd
+
+[codespell]
+skip = ./.git,./doc/themes/scikit-learn-modern/static/js,./sklearn/feature_extraction/_stop_words.py,./doc/_build,./doc/auto_examples,./doc/modules/generated
```
May we add `./.mypy_cache` here too?

DimitriPapadopoulos replied:
I have added `./.mypy_cache`.

Using two exclusion mechanisms is complicated and error-prone, I agree with that. However, I'm worried that some of the words that we ignore will shadow very common typos, the kind you see in any large project. The most striking example is `teh`.

DimitriPapadopoulos replied:
I was hoping that at least files `v*.rst` wouldn't change much.

thomasjpfan reviewed (commented):
I think `codespell_exclude.txt ` is a little too fragile. If there is any change in those lines, we may need to update `codespell_exclude.txt`. I am happy with the original version where we had the names & words in `codespell_ignore_words.txt`.

thomasjpfan reviewed (approved):
LGTM
